### PR TITLE
Fix the docs for supported keyword args or Rails 5 style integration controller tests [ci skip]

### LINKED ISSUE #25394: Can't set session in controller test http requests
### Steps to reproduce

I generated a new rails 5.0.0.rc1 app along with a home controller and a sessions controller:

``` ruby
    # in routes.rb
    get '/login', to: 'sessions#new', as: 'login'
    root to: 'home#show'

    # sessions controller
    class SessionsController < ApplicationController 
      def new
        if session[:user_id]
          redirect_to root_url, notice: 'Already logged in'
        end
      end 
    end
 ```

Now the first thing I notice is that my auto generated test file inherits from ActionDispatch::IntegrationTest:

``` ruby
    # test/controllers/sessions_controller_test.rb
    class SessionsControllerTest < ActionDispatch::IntegrationTest
    end    
```

So lets try to write a test where we set the `user_id` session key so we can ensure that our logged in user is redirected to the root path when they visit the login page. According to the [Edge Guide](http://edgeguides.rubyonrails.org/testing.html#functional-tests-for-your-controllers) we're supposed to use named arguments with HTTP requests in tests now:

``` ruby
    # test/controllers/sessions_controller_test.rb
    class SessionsControllerTest < ActionDispatch::IntegrationTest
      test 'should redirect to root path if a logged in user visits the login page' do
        get login_url, session: { user_id: 42 }
        assert_redirected_to root_url, 'Did not redirect'
      end
    end    
```

This test fails because the redirect is **not followed** and we also see the following deprecation message:

>DEPRECATION WARNING: ActionDispatch::IntegrationTest HTTP request methods will accept only
the following keyword arguments in future Rails versions:
params, headers, env, xhr

>Examples:

>get '/profile',
  params: { id: 1 },
  headers: { 'X-Extra-Header' => '123' },
  env: { 'action_dispatch.custom' => 'custom' },
  xhr: true

So rails is not letting us set the session in the way we are advised to in the Edge Guide.

It looks like the code that enables the use of a kw_arg named 'session' lives in [ActionController::TestCase](https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/test_case.rb#L462) and in order to fix this issue I had to do the following:

1 - Change SessionsControllerTest to inherit from ActionControllerTestCase:

``` ruby
    class SessionsControllerTest < ActionController::TestCase
```

2 - Use symbolized action names instead of URL's with http verbs

``` ruby
    # changed from 
    get '/login', session: { user_id: 42 }

   # to
   get :new, session: { user_id: 42 }
```

It's worth pointing out that we can't even set session variables manually when inheriting from ActionDispatch::IntegrationTest i.e:

``` ruby
    test 'some test' do
      session[:user_id] = 42
      assert something
    end
```
>NoMethodError: undefined method `session' for nil:NilClass

### Expected behavior
I was expecting to be able to pass keyword arguments to HTTP verbs in my controller tests like so:

    get '/some-url', session: { cart_id: 42 }

### Actual behavior
Keyword arguments appear to be ignored unless you inherit from ActionController::TestCase

### System configuration
**Rails version**:
5.0.0.rc1

**Ruby version**:
mri 2.3.0p0

### PR DESCRIPTION
**Pull request by prathamesh-sonpatki**
r? @eileencodes 
- Fixes #25394.

### CODE REVIEW

**Comment by prathamesh-sonpatki** [comment]
This needs backport too.

File: guides/source/testing.md [hunk 1]
```diff
@@ -695,24 +695,28 @@ end
...
 * `params`: option with a hash of request parameters to pass into the action
   (e.g. query string parameters or article variables).
 
-* `session`: option with a hash of session variables to pass along with the request.
+* `headers`: for setting the headers that will be passed with the request.
 
-* `flash`: option with a hash of flash values.
+* `env`: for customizing the request environment as needed.
+
+* `xhr`: weather the request is Ajax request or not. Can be set to true for marking the request as Ajax.
```
Comment (eileencodes) [inline guides/source/testing.md]: You have a typo here, should be "whether the request"

Comment (prathamesh-sonpatki) [inline guides/source/testing.md]: Oops. Fixed :)

File: guides/source/testing.md [hunk 2]
```diff
@@ -695,24 +695,28 @@ end
...
+* `as`: for encoding the request with different content type. Supports `:json` by default.
 
 All of these keyword arguments are optional.
 
-Example: Calling the `:show` action, passing an `id` of 12 as the `params` and setting a `user_id` of 5 in the session:
+Example: Calling the `:show` action, passing an `id` of 12 as the `params` and setting the request encoder as JSON:
 
 ```ruby
-get(:show, params: { id: 12 }, session: { user_id: 5 })
+get(:show, params: { id: 12 }), as: :json
```
Comment (kaspth) [inline guides/source/testing.md]: Closing paren is set wrong :)

File: guides/source/testing.md [hunk 3]
```diff
@@ -695,24 +695,28 @@ end
...
+* `env`: for customizing the request environment as needed.
+
+* `xhr`: whether the request is Ajax request or not. Can be set to true for marking the request as Ajax.
+
+* `as`: for encoding the request with different content type. Supports `:json` by default.
 
 All of these keyword arguments are optional.
 
-Example: Calling the `:show` action, passing an `id` of 12 as the `params` and setting a `user_id` of 5 in the session:
+Example: Calling the `:show` action, passing an `id` of 12 as the `params` and setting the request encoder as JSON:
```
Comment (kaspth) [inline guides/source/testing.md]: Request encoder is an internal term, would rather say "encodes the request as JSON".

File: guides/source/testing.md (hunk 2 above)
Comment (kaspth) [inline guides/source/testing.md]: How does the as option make sense on get requests?

Comment (prathamesh-sonpatki) [inline guides/source/testing.md]: @kaspth :+1: Updated example with headers.

File: guides/source/testing.md (hunk 3 above)
Comment (prathamesh-sonpatki) [inline guides/source/testing.md]: Updated the example

File: guides/source/testing.md [hunk 4]
```diff
@@ -695,30 +695,34 @@ end
 In the `test_should_get_index` test, Rails simulates a request on the action called `index`, making sure the request was successful
 and also ensuring that the right response body has been generated.
 
-The `get` method kicks off the web request and populates the results into the `@response`. It accepts 4 arguments:
+The `get` method kicks off the web request and populates the results into the `@response`. It can accept upto 6 arguments:
```
Comment (connorshea) [inline guides/source/testing.md]: "up to" should be two words

Comment (prathamesh-sonpatki) [inline guides/source/testing.md]: Thanks, updated

**Comment by kaspth** [comment]
Nice 👍

**Comment by prathamesh-sonpatki** [comment]
@kaspth We need to backport it too :)

**Comment by kaspth** [comment]
@prathamesh-sonpatki doh, nice catch! Backported to 5-0-stable @ 91f7779, 5-0-0 @ 662be0c 😁

**Comment by kaspth** [comment]
Ah, this didn't need to go in to 5-0-0. I was a little too quick there.

**Pull request merged**